src/main/java/org/jabref/logic/importer/fetcher/MathSciNet.java

```diff
@@ -102,34 +115,123 @@ public URL getUrlForIdentifier(String identifier) throws URISyntaxException, Mal
+                        JSONArray entriesArray = element.getJSONArray("results");
+                        for (int i = 0; i < entriesArray.length(); i++) {
+                            JSONObject entryObject = entriesArray.getJSONObject(i);
+                            BibEntry bibEntry = jsonItemToBibEntry(entryObject);
+                            entries.add(bibEntry);
+                        }
                     }
                 }
-            } catch (JSONException | TokenMgrException e) {
+            } catch (JSONException | ParseException e) {
                 LOGGER.error("An error occurred while parsing fetched data", e);
                 throw new ParseException("Error when parsing entry", e);
             }
             return entries;
         };
     }
 
+    private BibEntry jsonItemToBibEntry(JSONObject item) throws ParseException {
+        try {
+            BibEntry entry = new BibEntry(StandardEntryType.Article);
+            // Set fields based on the mappings
+            for (Map.Entry<StandardField, List<String>> mapEntry : fieldMappings.entrySet()) {
+                StandardField field = mapEntry.getKey();
+                List<String> path = mapEntry.getValue();
+
+                String value;
+                if (field == StandardField.AUTHOR) {
+                    value = toAuthors(item.optJSONArray(path.getFirst()));
+                } else if (field == StandardField.KEYWORDS) {
+                    value = getKeywords(item.optJSONObject(path.getFirst()));
+                } else {
+                    value = getOrNull(item, path).orElse(null);
+                }
+
+                if (value != null) {
+                    entry.setField(field, value);
+                }
+            }
+            // Handle articleUrl and mrnumber fields separately
+            String doi = item.optString("articleUrl", "");
+            if (!doi.isEmpty()) {
+                entry.setField(StandardField.DOI, doi);
+            }
+
+            String mrNumber = item.optString("mrnumber", "");
+            if (!mrNumber.isEmpty()) {
+                entry.setField(StandardField.MR_NUMBER, mrNumber);
+            }
+            return entry;
+        } catch (JSONException exception) {
+            throw new ParseException("MathSciNet API JSON format has changed", exception);
+        }
+    }
+
+    private Optional<String> getOrNull(JSONObject item, List<String> keys) {
+        Object value = item;
+        for (String key : keys) {
+            if (value instanceof JSONObject) {
+                value = ((JSONObject) value).opt(key);
+            } else if (value instanceof JSONArray) {
+                value = ((JSONArray) value).opt(Integer.parseInt(key));
+            } else {
+                break;
+            }
+        }
+
+        if (value instanceof String stringValue) {
+            return Optional.of(new String(stringValue.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8));
```

Comment:
Can you explain this necessity of charset conversion? Why can't the string be used directly?

Reply:
Yes, if used directly, I was getting these characters, for example:
![Screenshot 2024-03-20 004828](https://github.com/JabRef/jabref/assets/74734844/7f26c366-6d4e-47ec-8a8b-91f5e893d579)

Whereas the expected was Alonso-Morón, M. /Gómez, A.González. And many more.

Reply:
Please add a comment to a new method. That code appears twice: move it to a method `fixStringEncoding`

Reply:
Will do. Thank you for putting so much effort reviewing. I'll make all these changes an put another review request.